modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/CacheWorker.kt

```diff
@@ -0,0 +1,106 @@
+
+@OptIn(UnstableApi::class)
+@HiltWorker
+class CacheWorker @AssistedInject constructor(
+    @Assisted private val context: Context,
+    @Assisted params: WorkerParameters,
+    private val exoPlayerHelper: ExoPlayerHelper,
+) : Worker(context, params) {
+    private var cacheWriter: CacheWriter? = null
+
+    override fun doWork(): Result {
+        try {
+            val url = inputData.getString(URL_KEY)
+            val episodeUuid = inputData.getString(EPISODE_UUID_KEY)
+            val uri = Uri.parse(url)
+
+            val dataSourceFactory = exoPlayerHelper.getDataSourceFactory()
+            var dataSpec = DataSpec(uri)
+            if (episodeUuid != null) {
+                dataSpec = dataSpec.buildUpon().setKey(episodeUuid).build()
+            }
+
+            val cacheDataSourceFactory = CacheDataSource.Factory()
+                .setUpstreamDataSourceFactory(dataSourceFactory)
+                .setFlags(CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR)
+            exoPlayerHelper.getSimpleCache()?.let {
+                cacheDataSourceFactory.setCache(it)
+            }
+
+            cacheWriter = CacheWriter(
+                cacheDataSourceFactory.createDataSource(),
+                dataSpec,
+                null,
+            ) { requestLength: Long, bytesCached: Long, _: Long ->
+                if (bytesCached == requestLength) {
+                    Timber.d("$TAG: Caching complete for episode id: $episodeUuid")
+                }
+            }
+            cacheWriter?.cache()
+        } catch (exception: Exception) {
+            Timber.e("$TAG: ${exception.message}")
+        }
+        return Result.success()
+    }
+
+    override fun onStopped() {
+        try {
+            cacheWriter?.cancel()
+            super.onStopped()
+        } catch (exception: Exception) {
+            Timber.e("$TAG: ${exception.message}")
+        }
+    }
+
+    companion object {
+        private const val TAG = "CacheWorker"
+        private const val CACHE_WORKER_TAG = "pocket_casts_cache_worker_tag"
+        private const val URL_KEY = "url_key"
+        private const val EPISODE_UUID_KEY = "episode_uuid_key"
+
+        fun startCachingEntireEpisode(
+            context: Context,
+            url: String?,
+            episodeUuid: String?,
+        ) {
+            val inputData = Data.Builder()
+                .putString(URL_KEY, url)
+                .putString(EPISODE_UUID_KEY, episodeUuid)
+                .build()
+
+            // Cancel previous caching work
+            WorkManager.getInstance(context).cancelAllWorkByTag(CACHE_WORKER_TAG)
+            val constraints = Constraints.Builder()
+                .setRequiredNetworkType(NetworkType.UNMETERED)
+                .build()
+
+            val cacheWorkRequest = OneTimeWorkRequest.Builder(CacheWorker::class.java)
+                .addTag(CACHE_WORKER_TAG)
+                .setConstraints(constraints)
+                .setInputData(inputData).build()
```

Comment:
Should caching start if either episode UUID or url are `null`?

Reply:
Oh yes, I added an error in [4fe6dea](https://github.com/Automattic/pocket-casts-android/pull/2380/commits/4fe6dea2023d6063b51ee43f40353fa70538f99c)